### PR TITLE
adding CucumberRule to JUnit module

### PR DESCRIPTION
**Pull request by rmannibucau**
It allows to use RuleChain to build chain of rules and do things before/after cucumber itself reusing existing JUnit integration.

### CODE REVIEW

**Comment by haroon-sheikh** [comment]
This looks good to me.

**Comment by mariaklimenko** [comment]
Sounds useful

**Comment by haroon-sheikh** [comment]
@aslakhellesoy can this be merged in please?

File: junit/src/main/java/cucumber/api/junit/CucumberRule.java [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+                } finally {
+                    jUnitReporter.done();
+                    jUnitReporter.close();
+                    runtime.printSummary();
+                }
+            }
+        };
+    }
+
+    public static class ForceRunner extends BlockJUnit4ClassRunner {
```
Comment (mpkorstanje) [inline junit/src/main/java/cucumber/api/junit/CucumberRule.java]: What is the purpose of this?

File: junit/src/main/java/cucumber/api/junit/CucumberRule.java [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+                try {
+                    for (final CucumberFeature feature : cucumberFeatures) {
+                        new FeatureRunner(feature, runtime, jUnitReporter).run(new RunNotifier());
+                    }
+                    base.evaluate();
+                } finally {
+                    jUnitReporter.done();
+                    jUnitReporter.close();
+                    runtime.printSummary();
+                }
```
Comment (mpkorstanje) [inline junit/src/main/java/cucumber/api/junit/CucumberRule.java]: This won't work any more. We're using an event based protocol now.

File: junit/src/main/java/cucumber/api/junit/CucumberRule.java [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+                options.add(MultiLoader.CLASSPATH_SCHEME + defaultPackage.replace('.', '/'));
+            }
+            if (!hasOption(options, "--glue")) {
+                options.add("--glue");
+                options.add(defaultPackage);
+            }
+            if (!hasOption(options, "--plugin")) {
+                options.add("--plugin");
+                options.add("pretty");
+            }
```
Comment (mpkorstanje) [inline junit/src/main/java/cucumber/api/junit/CucumberRule.java]: I reckon a builder pattern would be more suited to pass options.

File: junit/src/main/java/cucumber/api/junit/CucumberRule.java [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+import java.util.Collections;
+import java.util.List;
+
+import static java.util.Arrays.asList;
+
+public class CucumberRule implements TestRule {
+    private final RuntimeOptions runtimeOptions;
+
+    public CucumberRule() {
+        this(new RuntimeOptionsFactory(CucumberRule.class /* no conf */).create());
```
Comment (mpkorstanje) [inline junit/src/main/java/cucumber/api/junit/CucumberRule.java]: This only makes sense for classes that are annotated with @CucumberOptions

File: junit/src/main/java/cucumber/api/junit/CucumberRule.java (hunk 3 above)
Comment (rmannibucau) [inline junit/src/main/java/cucumber/api/junit/CucumberRule.java]: Agree but less cucumber like. Maybe a core refactoring to build cli on top of it?

File: junit/src/main/java/cucumber/api/junit/CucumberRule.java (hunk 1 above)
Comment (rmannibucau) [inline junit/src/main/java/cucumber/api/junit/CucumberRule.java]: Make junit happy without loosing stand runner features. Can be replaced with a more advanced/complete custom runner but harder to maintain probably.

File: junit/src/main/java/cucumber/api/junit/CucumberRule.java (hunk 4 above)
Comment (rmannibucau) [inline junit/src/main/java/cucumber/api/junit/CucumberRule.java]: Not sure what you mean. Goal of the rule was to comibe cucumber with other rules (container, ftp/sft/... server, mysql, nosql databaseq etc...)

**Comment by mpkorstanje** [comment]
>Make junit happy without loosing stand runner features. Can be replaced with a more advanced/complete custom runner but harder to maintain probably.

I'm not happy with the ForceRunner or NoTestRunner. It is essentially a new runner. And having looked at JUnits rules more closely I now understand that this is an alteration _in how a test method, or set of test methods, is run and reported_. So a Cucumber Rule that executes the test goes against the design framework. This makes both Cucumber and Junit harder to use.

So I believe the only correct solution is to implement rules an annotation in the Cucumber Runner. However the current implementation already extends the `ParentRunner` which should be able handle `@ClassRules`. So it is already possible to use rules to do things around the execution of Cucumber.

So I'm not sure what use case this solution actually has.

**Comment by rmannibucau** [comment]
This reasonning is partially valid - it ignores rules ordering which is important - but add to it you can use another runner (arquillian, applicationcomposer, spring, apache directory, ...) and rules still make sense.

Also note this small workaround avoids a new api which is not bad.

We can probably mix cucumber rule ans runner to reuse existing one if needed bit making explicit it is a junit passthrough impl is not that blocking imo.

**Comment by mpkorstanje** [comment]
Unless I'm mistaken the `RuleChain` can be used in `@ClassRule` so there should be sufficient ability to order the test rules. However they have to be ordered **around** `Cucumber`. 

It would be nice if multiple runners could contribute but this is not something that JUnit4 supports. JUnit5 Jupiter does support these things - though it was still in its conceptional phase when this PR was created this is no longer the case.

As such I see no good reason to work around JUnit4s limitations or support such a workaround in the long term as part of cucumber-jvm. Finishing Junit5 support and upgrading to JUnit5 would be the most sensible and sustainable longterm approach.

**Comment by rmannibucau** [comment]
@mpkorstanje multiple runners = rule chain actually. JUnit 5 supports it but it is also not the next version of JUnit 4 since it breaks all the API and support of JUnit 4 (it is a bit like saying "TestNG supports it"). So still think it makes sense - mainly cause there is a huge ecosystem here already - to support rules.

**Comment by mpkorstanje** [comment]
> So still think it makes sense - mainly cause there is a huge ecosystem here already - to support rules.

Cucumber currently supports rules. And this is not what I understand your use case to be. I understand your use case to be integration with multiple runners. This is not something supported by Junit4. So I do not wish to support a workaround in the long term as part of cucumber-jvm.

You are of course free to create your own module that wraps the `Cucumber` runner in rules or any other sort of workaround.

**Comment by rmannibucau** [comment]
@mpkorstanje hmm, JUnit 4 supports chain of runners...this is why rules were created actually. Concretely it is not "runners" but "rules" but it leads to the same outcome, just more powerful. I pushed an example at https://github.com/rmannibucau/rule-them-all#user-content-combine-them. Cucumber would be welcomed being the last rule in the sample and we could append a cukedoctor + asciidoctor rendering rule after cucumber itself so I really think it makes a lot of sense to have cucumber as a rule. What am I missing?

**Pull request closed without merging**